### ISSUE TITLE
Jepsen tests are incomplete because they cannot run due to Java Runtime exceptions

### ISSUE THREAD

**Issue by mangalaman93** [dgraph-io/dgraph#3902]

**Comment by martinmr** [comment]
Fixed by https://github.com/jepsen-io/jepsen/pull/411

**Comment by martinmr** [comment]
Jepsen has been updated to use the newest API. Issue is being kept open due to some failures seen while running the tests (unrelated to the API changes).

**Comment by campoy** [comment]
Please add more information about those tests failing here.
This is very urgent.

**Comment by mangalaman93** [comment]
@prashant-shahi please post the details of running the tests here. I will look into this first thing Monday.

**Comment by martinmr** [comment]
Assigning to @prashant-shahi since I looked at it when this issue was about making Jepsen work with the API changes.

**Comment by prashant-shahi** [comment]
@mangalaman93  @martinmr  @campoy Right now, 6 of the move-tablet Jepsen tests are failing. In case of the incomplete tests if re-run (From 1st to 6th attempt), we are able to pass it.

Here is the complete report of the same.
https://discuss.dgraph.io/t/jepsen-tests-violations-and-exceptions/4892/7?u=prashant-shahi

**Comment by manishrjain** [comment]
Feel free to assign them to me.

**Comment by martinmr** [comment]
Spent most of the day thinking the disconnected span issue was due to incompatible versions. I managed to build a version of the current java client with the same versions of grpc and opencensus of the last known working version of jepsen but the issue is still there.

Once that approach failed, I tried going upward from version 1.6.0 and try to find the point at which the client stops working. This happens between versions 1.7.1 and 1.7.3. After more testing, I found the first PR that breaks: https://github.com/dgraph-io/dgraph4j/pull/80

Running jepsen with a version that includes that PR as the last change returns the following errors.

```
java.lang.RuntimeException: The commit encountered an execution exception:
        at io.dgraph.DgraphAsyncClient.lambda$runWithRetries$2(DgraphAsyncClient.java:195) ~[dgraph4j-1.7.2.jar:na]
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590) ~[na:1.8.0_222]
        at java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1582) ~[na:1.8.0_222]
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) ~[na:1.8.0_222]
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056) ~[na:1.8.0_222]
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692) ~[na:1.8.0_222]
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157) ~[na:1.8.0_222]
Caused by: java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: ABORTED: Transaction has been aborted. Please retry.
        at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357) ~[na:1.8.0_222]
        at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895) ~[na:1.8.0_222]
        at io.dgraph.DgraphAsyncClient.lambda$runWithRetries$2(DgraphAsyncClient.java:173) ~[dgraph4j-1.7.2.jar:na]
        ... 6 common frames omitted
Caused by: io.grpc.StatusRuntimeException: ABORTED: Transaction has been aborted. Please retry.
        at io.grpc.Status.asRuntimeException(Status.java:526) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:434) ~[grpc-stub-1.15.0.jar:1.15.0]
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:678) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:403) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:459) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:63) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:546) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:467) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:584) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123) ~[grpc-core-1.15.0.jar:1.15.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_222]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_222]
        at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_222]
ERROR [2019-10-10 01:08:31,400] main - jepsen.cli Oh jeez, I'm sorry, Jepsen broke. Here's why:
java.lang.RuntimeException: The commit encountered an execution exception:
        at io.dgraph.DgraphAsyncClient.lambda$runWithRetries$2(DgraphAsyncClient.java:195) ~[dgraph4j-1.7.2.jar:na]
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590) ~[na:1.8.0_222]
        at java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1582) ~[na:1.8.0_222]
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) ~[na:1.8.0_222]
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056) ~[na:1.8.0_222]
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692) ~[na:1.8.0_222]
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157) ~[na:1.8.0_222]
Caused by: java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: ABORTED: Transaction has been aborted. Please retry.
        at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357) ~[na:1.8.0_222]
        at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895) ~[na:1.8.0_222]
        at io.dgraph.DgraphAsyncClient.lambda$runWithRetries$2(DgraphAsyncClient.java:173) ~[dgraph4j-1.7.2.jar:na]
        ... 6 common frames omitted
Caused by: io.grpc.StatusRuntimeException: ABORTED: Transaction has been aborted. Please retry.
        at io.grpc.Status.asRuntimeException(Status.java:526) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:434) ~[grpc-stub-1.15.0.jar:1.15.0]
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:678) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:403) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:459) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:63) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:546) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:467) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:584) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) ~[grpc-core-1.15.0.jar:1.15.0]
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123) ~[grpc-core-1.15.0.jar:1.15.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_222]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_222]
        at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_222]
Welcome to Jepsen on Docker
===========================
```

Something about making the client asynchronous broke jepsen and might also be responsible for the disconnected spans.

@mangalaman93: I'll keep working on this tomorrow but could you look at that PR and see if there's something obvious that would cause opencensus to break. Thanks.

**Comment by martinmr** [comment]
I figured out the OpenCensus problem. The fix is at https://github.com/dgraph-io/dgraph4j/pull/117.

**Comment by aphyr** [comment]
I've tracked down the issue that caused the Jepsen tests to crash so frequently: the exceptions thrown by the new Java client have a different structure. Instead of throwing just a io.grpc.StatusRuntimeException, it now throws things like a RuntimeException wrapping a j.u.c.CompletionException wrapping a RuntimeException wrapping a j.u.c.ExectionException wrapping a io.grpc.StatusRuntimeException. Since the exception classes changed, try/catch blocks no longer caught errors properly. I imagine this change probably broke error handling for everyone who used the Java client.

Dgraph can be a bit fragile on setup, and Jepsen relies heavily on catching exceptions and retrying operations to set up the cluster, initialize schemata, and create initial records. The change in exception structure bypassed those recovery mechanisms, which meant Jepsen tests would fail *immediately* when Dgraph started returning errors. That's why the tests became so fragile after the client change!

I've added code to unwrap this exception hierarchy to Jepsen, and updated all the critical paths I could find in the test suite. There'll be more for me to incrementally polish, but Jepsen 3bff032adf3a4277e5cbbc2cd05ecec90c69f61e is a *lot* more reliable now. :)
